### PR TITLE
Add DeblurGAN-v2 model

### PR DESCRIPTION
**Pull request by akorobeinikov**
TODO:

- [x] Add GFLOPs and MParams
- [x] Select only one model for submission

Issues:

- [x] Is it possible to convert Batch Norm from PyTorch to ONNX in train mode without decomposition this operation?

- [x] Does decomposition affect the speed of the model?

### CODE REVIEW

File: models/public/deblurgan-v2/accuracy-check.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            scale_border: 0
+            presenter: print_vector
+          - type: ssim
+            presenter: print_vector
+
+  - name: deblurgan-v2
+
+    launchers:
+      - framework: dlsdk
+        model: deblurgan-v2.xml
```
Comment (eaidova) [inline models/public/deblurgan-v2/accuracy-check.yml:39]: please do not forget to remove path to the model and aligh name with original

File: tools/accuracy_checker/accuracy_checker/postprocessor/corner_crop.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+"""
+
+from ..config import NumberField, StringField
+from ..utils import get_size_from_config
+from .postprocessor import PostprocessorWithSpecificTargets
+from ..preprocessor import CornerCrop
+from ..representation import ImageProcessingAnnotation, ImageProcessingPrediction
+
+
+class CornerCropImage(PostprocessorWithSpecificTargets):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/corner_crop.py:24]: if I right remember, we already have simmilar postprocessing:
https://github.com/openvinotoolkit/open_model_zoo/blob/develop/tools/accuracy_checker/accuracy_checker/postprocessor/crop_ground_truth_image.py#L58
can we merge them into 1?

**Comment by akorobeinikov** [comment]
Jenkins please retry a build

File: models/public/deblurgan-v2/deblurgan-v2.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `RGB`. Mean values - [127.5,127.5,127.5], scale values - [127.5,127.5,127.5].
+
+### Converted Model
+
+Image, name - `blur_image`,  shape - `[1x3x736x1312]`, format is `[BxCxHxW]`, where:
```
Comment (eizamaliev) [inline models/public/deblurgan-v2/deblurgan-v2.md:46]:
```suggestion
Image, name - `blur_image`,  shape - `1,3,736,1312`, format is `B,C,H,W`, where:
```

File: models/public/deblurgan-v2/deblurgan-v2.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+| Metric | Original model | Converted model |
+| ------ | -------------- | --------------- |
+| PSNR   | 28.25Db        | 28.25Db         |
+| SSIM   | 0.97           | 0.97            |
+
+## Input
+
+### Original Model
+
+Image, name - `blur_image`,  shape - `[1x3x736x1312]`, format is `[BxCxHxW]`, where:
```
Comment (eizamaliev) [inline models/public/deblurgan-v2/deblurgan-v2.md:35]:
```suggestion
Image, name - `blur_image`,  shape - `1,3,736,1312`, format is `B,C,H,W`, where:
```

**Review by eizamaliev — CHANGES_REQUESTED** [review]

**Comment by eaidova** [comment]
@akorobeinikov please rebase your changes

File: demos/python_demos/deblurring_demo/model.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                      'constant_values': 0,
+                      'pad_width': ((0, self.h - image.shape[0]), (0, self.w - image.shape[1]), (0, 0))
+                      }
+
+        if image.shape[0] != self.h or image.shape[1] != self.w:
+            resized_image = self._padding(normalized_image, pad_params)
+        else:
+            resized_image = normalized_image
+        resized_image = resized_image.transpose((2, 0, 1))
+        resized_image = resized_image.reshape((self.n, self.c, self.h, self.w))
```
Comment (eaidova) [inline demos/python_demos/deblurring_demo/model.py:104]: better to use np.expand_dims instead

File: models/public/deblurgan-v2/deblurgan-v2.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+### Converted Model
+
+Deblurred image, name - `deblur_image`,  shape - `1,3,736,1312`, output data format is `B,C,H,W` where:
+
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `RGB`.
```
Comment (eaidova) [inline models/public/deblurgan-v2/deblurgan-v2.md:77]: why is it still RGB?
if normalization is not performed during MO conversion, it should be noted in this section

File: tools/accuracy_checker/dataset_definitions.yml [hunk 7]
```diff
@@ -1142,3 +1142,11 @@ datasets:
...
       input_suffix: in
       reference_suffix: out
+
+  - name: GoPro
+    data_source: GOPRO/test
+    annotation_conversion:
+      converter: image_processing
+      data_dir: GOPRO/test
+      input_suffix: blur.png
+      target_suffix: sharp.png
```
Comment (eaidova) [inline tools/accuracy_checker/dataset_definitions.yml:1152]: please add annotation pickle (required for testing)

File: models/public/deblurgan-v2/accuracy-check.yml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            presenter: print_vector
+
+  - name: deblurgan-v2
+
+    launchers:
+      - framework: dlsdk
+        adapter:
+          type: image_processing
+          mean: 127.5
+          std: 127.5
```
Comment (eaidova) [inline models/public/deblurgan-v2/accuracy-check.yml:41]: please note that PSNR sensitive to channels order, you need to align output image channels order (BGR) with gt channels order (RGB) using `reverse_channels: True` in adapter

File: demos/python_demos/deblurring_demo/model.py (hunk 5 above)
Comment (akorobeinikov) [reply demos/python_demos/deblurring_demo/model.py]: done

File: tools/accuracy_checker/dataset_definitions.yml (hunk 7 above)
Comment (akorobeinikov) [reply tools/accuracy_checker/dataset_definitions.yml]: done

File: models/public/deblurgan-v2/deblurgan-v2.md (hunk 6 above)
Comment (akorobeinikov) [reply models/public/deblurgan-v2/deblurgan-v2.md:77]: Output in `RGB` format as in original model. For IR model using reverse_input_channels.

File: models/public/deblurgan-v2/accuracy-check.yml (hunk 8 above)
Comment (akorobeinikov) [reply models/public/deblurgan-v2/accuracy-check.yml:41]: Parameters of adapter are right, output image channels order is RGB.

**Comment by vladimir-dudnik** [comment]
@akorobeinikov please resolve conflicts

**Comment by eaidova** [comment]
Jenkins please retry a build

File: models/public/deblurgan-v2/model.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+        parameters = {'g_name': model_name, 'norm_layer': 'instance'}
+        self.impl = get_generator(parameters)
+        checkpoint = torch.load(weights, map_location='cpu')['model']
+        self.impl.load_state_dict(checkpoint)
+        self.impl.train(True)
+        remove_all_batch_norm(self.impl)
+
+    def forward(self, image):
+        return self.impl(image)
```
Comment (eaidova) [inline models/public/deblurgan-v2/model.py:57]: @akorobeinikov do you have plan align model behavior with other models (insert postprocessing - normalize output to [0, 1] range and change channels order?
Possibly it will be need for demo unification

Comment (akorobeinikov) [reply models/public/deblurgan-v2/model.py]: I did dome changes to unify: normalize output to [0, 1] range and change channels order to BGR. Also I make preprocessing normalize as part of mo conversion.

File: models/public/deblurgan-v2/deblurgan-v2.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+For details about implementation of model,
+check out the [DeblurGAN-v2: Deblurring (Orders-of-Magnitude) Faster and Better](https://github.com/VITA-Group/DeblurGANv2) repository.
+
+## Specification
+
+GFLOPs and MParams are presented for Mobilenet backbone.
+
+| Metric            | Value           |
+|-------------------|-----------------|
+| Type              | ImageProcessing |
```
Comment (eaidova) [inline models/public/deblurgan-v2/deblurgan-v2.md:17]:
```suggestion
| Type              | Image Processing |
```

File: models/public/deblurgan-v2/deblurgan-v2.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+DeblurGAN-v2 is a generative adversarial network (GAN) for single image motion deblurring.
+This model is based on a relativistic conditional GAN with a double-scale discriminator.
+For details about architecture of model, check out the [paper](https://arxiv.org/abs/1908.03826).
+Model used Mobilenet as backbone and was trained on GoPro, DVD, NFS datasets.
+For details about implementation of model,
+check out the [DeblurGAN-v2: Deblurring (Orders-of-Magnitude) Faster and Better](https://github.com/VITA-Group/DeblurGANv2) repository.
+
+## Specification
+
+GFLOPs and MParams are presented for Mobilenet backbone.
```
Comment (eaidova) [inline models/public/deblurgan-v2/deblurgan-v2.md:13]: not sure that we need such comment if we have only one model in OMZ

File: models/public/deblurgan-v2/model.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        self.impl = get_generator(parameters)
+        checkpoint = torch.load(weights, map_location='cpu')['model']
+        self.impl.load_state_dict(checkpoint)
+        self.impl.train(True)
+        remove_all_batch_norm(self.impl)
+
+    def forward(self, image):
+        out = self.impl(image)
+        # convert out to [0, 1] range and change channel order RGB->BGR
+        out = (out + 1)/2
```
Comment (eaidova) [inline models/public/deblurgan-v2/model.py:59]:
```suggestion
        out = (out + 1) / 2
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/crop_image.py [hunk 13]
```diff
@@ -45,31 +50,49 @@ def parameters(cls):
...
     def process_image(self, annotation, prediction):
+        target_height = self.dst_height or self.image_size[0]
+        target_width = self.dst_width or self.image_size[1]
+
+        for target in annotation:
+            target.value = Crop.process_data(
+                target.value, target_height, target_width, None, False, False, True, {}
+            )
+
         for target in annotation:
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/crop_image.py:64]: why do you crop annotations twice???

File: models/public/deblurgan-v2/model.py (hunk 9 above)
Comment (eaidova) [reply models/public/deblurgan-v2/model.py]: ok, thanks!

Also we need to update onnx model in internal repo after these changes

File: models/public/deblurgan-v2/deblurgan-v2.md (hunk 11 above)
Comment (akorobeinikov) [reply models/public/deblurgan-v2/deblurgan-v2.md]: Deleted

File: tools/accuracy_checker/accuracy_checker/postprocessor/crop_image.py (hunk 13 above)
Comment (akorobeinikov) [reply tools/accuracy_checker/accuracy_checker/postprocessor/crop_image.py]: Fixed

File: models/public/deblurgan-v2/accuracy-check.yml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  - name: deblurgan-v2
+
+    launchers:
+      - framework: dlsdk
+        adapter:
+          type: image_processing
+          reverse_channels: True
+
+    datasets:
+      - name: GOPRO
```
Comment (eaidova) [inline models/public/deblurgan-v2/accuracy-check.yml:42]:
```suggestion
      - name: GoPro
```

File: data/dataset_definitions.yml [hunk 15]
```diff
@@ -1167,3 +1167,12 @@ datasets:
       converter: place_recognition
       split_file: pitts250k/datasets/pitts30k_val.mat
     annotation: pitts30k_val.pickle
+
+  - name: GOPRO
```
Comment (eaidova) [inline data/dataset_definitions.yml:1171]:
```suggestion
  - name: GoPro
```

**Review by eaidova — APPROVED** [review]

**Comment by eaidova** [comment]
Jenkins please retry a build

File: demos/common/python/models/__init__.py [hunk 16]
```diff
@@ -27,6 +28,7 @@
     'CenterNet',
     'DetectionWithLandmarks',
     'FaceBoxes',
+    'Model',
```
Comment (Wovchena) [inline demos/common/python/models/__init__.py:31]: Put `DeblurringModel` into common folder instead.

File: demos/deblurring_demo/python/README.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+## How It Works
+
+Upon the start-up the demo application reads command line parameters and loads a network. The demo runs inference and shows results for each image captured from an input. Depending on number of inference requests processing simultaneously (-nireq parameter) the pipeline might minimize the time required to process each single image (for nireq 1) or maximize utilization of the device and overall processing performance.
+
+For each image demo performs the following steps:
+
+1. Do preprocessing consisting of normalization and padding to input shape of model.
+2. Inference of model (user is able to set the inference options to influence the execution process).
+3. Do postprocessing for output of model.
+4. To display the resulting image together with source image.
```
Comment (Wovchena) [inline demos/deblurring_demo/python/README.md:16]:
```suggestion
4. Display the resulting image together with source image.
```

File: demos/deblurring_demo/python/deblurring_demo.py [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+                            default=1, type=int)
+    infer_args.add_argument('-nstreams', '--num_streams',
+                            help='Optional. Number of streams to use for inference on the CPU or/and GPU in throughput '
+                                 'mode (for HETERO and MULTI device cases use format '
+                                 '<device1>:<nstreams1>,<device2>:<nstreams2> or just <nstreams>).',
+                            default='', type=str)
+    infer_args.add_argument('-nthreads', '--num_threads', default=None, type=int,
+                            help='Optional. Number of threads to use for inference on CPU (including HETERO cases).')
+
+    io_args = parser.add_argument_group('Input/output options')
```
Comment (Wovchena) [inline demos/deblurring_demo/python/deblurring_demo.py:63]: Please add `--output` and `--output_limit` arguments.

File: demos/deblurring_demo/python/model.py [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+            raise Exception("Unexpected output blob shape {}. Only 4D output blob is supported".format(output_size))
+
+        return output_blob_name
+
+    def preprocess(self, inputs):
+        image = inputs
+        # right bottom padding to resize input image to input_layer shape
+        pad_params = {'mode': 'constant',
+                      'constant_values': 0,
+                      'pad_width': ((0, self.h - image.shape[0]), (0, self.w - image.shape[1]), (0, 0))
```
Comment (Wovchena) [inline demos/deblurring_demo/python/model.py:83]: Second image can be of different size compared to the first one. This may result in `pad_width` having negative values. Raise an exception with meaningful error or use cv2.resize() to `self.h, self.w`.

File: demos/deblurring_demo/python/model.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+    def postprocess(self, outputs, meta):
+        prediction = outputs[self.output_blob_name].squeeze()
+        input_image_height = meta['original_shape'][0]
+        input_image_width = meta['original_shape'][1]
+
+        prediction = prediction.transpose((1, 2, 0))
+        prediction *= 255
+        prediction = prediction.astype(np.uint8)
+
+        return prediction[:input_image_height, :input_image_width, :]
```
Comment (Wovchena) [inline demos/deblurring_demo/python/model.py:107]: I think you should crop the image at first to minimize complexity of multiplying by 255 and type conversion.

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/deblurring_demo/python/deblurring_demo.py (hunk 18 above)
Comment (akorobeinikov) [reply demos/deblurring_demo/python/deblurring_demo.py:63]: Added

**Comment by eaidova** [comment]
Jenkins please retry a build

File: demos/common/python/models/deblurring.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        prediction = outputs[self.output_blob_name].squeeze()
+        input_image_height = meta['original_shape'][0]
+        input_image_width = meta['original_shape'][1]
+
+        prediction = prediction.transpose((1, 2, 0))
+        prediction = prediction[:min(self.h, input_image_height), :min(self.w, input_image_width), :]
+        prediction *= 255
+        prediction = prediction.astype(np.uint8)
+
+        return prediction
```
Comment (Wovchena) [inline demos/common/python/models/deblurring.py:104]:
```suggestion
        return prediction.astype(np.uint8)
```

File: demos/deblurring_demo/python/model.py (hunk 19 above)
Comment (Wovchena) [reply demos/deblurring_demo/python/model.py]: I believe it's still aplicable.

Comment (akorobeinikov) [reply demos/deblurring_demo/python/model.py]: Fixed

File: demos/common/python/models/deblurring.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+ limitations under the License.
+"""
+
+import cv2
+import math
+import numpy as np
+
+from .model import Model
+
+class DeblurringModel(Model):
```
Comment (Wovchena) [inline demos/common/python/models/deblurring.py:20]: You should follow @vladimir-dudnik's comment about the name.

Comment (akorobeinikov) [reply demos/common/python/models/deblurring.py]: Ok, I changed the name. But what about `SegmentationModel` and `DetectionWithLandmarks`?

**Review by Wovchena — APPROVED** [review]

File: demos/common/python/models/deblurring.py (hunk 22 above)
Comment (Wovchena) [reply demos/common/python/models/deblurring.py]: They are out of this PR scope any way.

**Comment by eaidova** [comment]
@eizamaliev please take a look again, your requested changes were adressed

**Review by eizamaliev — APPROVED** [review]
